docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,537 @@
+17.6.1, and 17.7). Each of these sections defines the scope (range of code) in
+which the variable's type is to be promoted.
+
+Type promotion for a variable v is allowed only when we can deduce that such
+promotion is valid based on an analysis of certain boolean expressions. In such
+cases, we either say that the boolean expression b shows that v has type T when
+b is true, or that the boolean expression b shows that v has type T when b is
+false. As a rule, for all variables v and types T, a boolean expression does not
+show that v has type T. Those situations where an expression does show that a
+variable has a type are mentioned explicitly in the relevant sections of this
+specification (16.34 and 16.22).
+
+Dart 1.0: Type promotion for a variable v is also allowed only for a subset of
+variables and is dependent on the use of the variable in the scope in which the
+promotion would occur. If a variable is one for which type promotion is allowed,
+we say that the variable is promotable in some scope. A variable v is promotable
+in the scope S if all of the following conditions are met:
+
+*   v is either a local variable or a parameter,
+
+*   v cannot be potentially mutated within any closure unless all of the types
+    of the values that would be assigned to v within S are more specific than
+    the promoted type of v, and
+
+*   if v is captured by a closure in S then v cannot be potentially mutated
+    anywhere unless all of the types of the values that would be assigned to v
+    within S are more specific than the promoted type of v.
+
+Dart 2.0: Type promotion for a variable v is also allowed only for a subset of
+variables and is dependent on the use of the variable in the scope in which the
+promotion would occur. If a variable is one for which type promotion is allowed,
+we say that the variable is promotable in some scope. A variable v is promotable
+in the scope S if all of the following conditions are met:
+
+*   v is either a local variable or a parameter,
+
+*   v cannot be potentially mutated within S unless all of the types of the
+    values that would be assigned to v within S are a subtype of the promoted
+    type of v,
+
+*   v cannot be potentially mutated within any closure unless all of the types
+    of the values that would be assigned to v within S are a subtype of the
+    promoted type of v, and
+
+*   if v is captured by a closure in S then v cannot be potentially mutated
+    anywhere unless all of the types of the values that would be assigned to v
+    within S are a subtype of the promoted type of v.
+
+For the purposes of type promotion, a statement cannot complete normally if the
+statement is either
+
+*   a return statement,
+*   a rethrow statement,
+*   an expression statement whose expression is a throw expression,
+*   an if statement of the form if (b) s1 else s2 where neither s1 nor s2 can
+    complete normally, or
+*   a block whose last statement cannot complete normally.
```

Comment:
I chose not to add it to the original proposal because (a) I wanted to keep this as simple as possible, (b) I don't think that case occurs frequently, and (c) it isn't necessary to add all possible enhancements at this point.

I'd leave it out for now, unless adding it will allow us to get this approved more quickly :-)